node/network/gossip-support/src/lib.rs

```diff
@@ -157,6 +168,80 @@ pub async fn connect_to_authorities(
 	failed_rx
 }
 
+/// We partition the list of all sorted `authorities` into sqrt(len) groups of sqrt(len) size
+/// and form a matrix where each validator is connected to all validators in its row and column.
+/// This is similar to [web3] research proposed topology, except for the groups are not parachain
+/// groups (because not all validators are parachain validators and the group size is small),
+/// but formed randomly via BABE randomness from two epochs ago.
+/// This limits the amount of gossip peers to 2 * sqrt(len) and ensures the diameter of 2.
+///
+/// [web3]: https://research.web3.foundation/en/latest/polkadot/networking/3-avail-valid.html#topology
+async fn update_gossip_topology(
+	ctx: &mut impl SubsystemContext,
+	our_index: usize,
+	authorities: Vec<AuthorityDiscoveryId>,
+	relay_parent: Hash,
+) -> Result<(), util::Error> {
+	// retrieve BABE randomness
+	let random_seed = {
+		let (tx, rx) = oneshot::channel();
+
+		ctx.send_message(RuntimeApiMessage::Request(
+			relay_parent,
+			RuntimeApiRequest::CurrentBabeEpoch(tx),
+		).into()).await;
+
+		let randomness = rx.await??.randomness;
+		let mut subject = [0u8; 40];
+		subject[..8].copy_from_slice(b"gossipsu");
+		subject[8..].copy_from_slice(&randomness);
+		sp_core::blake2_256(&subject)
+	};
+
+	// shuffle the indices
+	let mut rng: ChaCha20Rng = SeedableRng::from_seed(random_seed);
+	let len = authorities.len();
+	let mut indices: Vec<usize> = (0..len).collect();
+	indices.shuffle(&mut rng);
+	let mut reverse_indices = vec![0usize; len];
+	for (i, idx) in indices.iter().enumerate() {
+		reverse_indices[*idx] = i;
+	}
+
+	let neighbors = matrix_neighbors(reverse_indices[our_index], len);
```

Comment:
nit: it seems `reverse_indices` doesn't need to be fully computed, only `our_reverse_index`